### PR TITLE
Maintain modifiers on `Omit`

### LINKED ISSUE #31190: `Omit` from #31115 makes all properties required
Since the changes of #31115, `Omit<T, K>` makes all properties of `T` required.

***

**TypeScript Version:**  3.5.0-dev.20190501, 3.4.5

**Search Terms:** omit optional

**Code**

```ts
type A = {
    a: number;
    b?: string;
};

type B = Omit<A, 'a'>;
```

**Expected behavior:** `B` is `{ b?: string }` (`{ b?: string | undefined }` with `strictNullChecks`).

**Actual behavior:** `B` is `{ b: string }` (`{ b: string | undefined }` with `strictNullChecks`), with property `b` required.

**Playground Link:**  [link](https://www.typescriptlang.org/play/index.html#src=type%20Omit1%3CT%2C%20K%20extends%20keyof%20any%3E%20%3D%20Pick%3CT%2C%20Exclude%3Ckeyof%20T%2C%20K%3E%3E%3B%0D%0Atype%20Omit2%3CT%2C%20K%20extends%20keyof%20any%3E%20%3D%20%7B%0D%0A%20%20%20%20%5BP%20in%20Exclude%3Ckeyof%20T%2C%20K%3E%5D%3A%20T%5BP%5D%0D%0A%7D%3B%0D%0A%0D%0Atype%20A%20%3D%20%7B%0D%0A%20%20%20%20a%3A%20number%3B%0D%0A%20%20%20%20b%3F%3A%20string%3B%0D%0A%7D%3B%0D%0A%0D%0Atype%20B1%20%3D%20Omit1%3CA%2C%20'a'%3E%3B%0D%0A%2F%2F%20B1%20%3D%20%7B%20b%3F%3A%20string%3B%20%7D%0D%0A%0D%0Atype%20B2%20%3D%20Omit2%3CA%2C%20'a'%3E%3B%0D%0A%2F%2F%20B2%20%3D%20%7B%20b%3A%20string%3B%20%7D%0D%0A)

pin @DanielRosenwasser

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Fixes #31190.

### CODE REVIEW

**Comment by InExtremaRes** [comment]
It was nice to have a better error message with the name `Omit` in it... I suppose there is no way to have both :sweat_smile:

BTW, why `{ [P in Exclude<keyof T, K>]: T[P] }` doesn't preserve modifiers and `Pick` does?

File: src/lib/es5.d.ts [hunk 1]
```diff
@@ -1446,9 +1446,7 @@ type Extract<T, U> = T extends U ? T : never;
 /**
  * Construct a type with the properties of T except for those in type K.
  */
-type Omit<T, K extends keyof any> = {
```
Comment (jwbay) [inline src/lib/es5.d.ts]: 🤔Why can't this be `K extends keyof T` to make the type homomorphic?

Comment (DanielRosenwasser) [reply src/lib/es5.d.ts]: I don't believe that would make the type homomorphic, would it @ahejlsberg?

**Comment by DanielRosenwasser** [comment]
> BTW, why `{ [P in Exclude<keyof T, K>]: T[P] }` doesn't preserve modifiers and `Pick` does?

What it comes down to is just the rules for when a mapped type decides it should maintain the modifiers, and right now it never does anything "smart" if the constraint (the thing after the `in` keyword is a conditional type.

> It was nice to have a better error message with the name `Omit` in it... 

Yeah. I agree. ☹

File: src/lib/es5.d.ts (hunk 1 above)
Comment (DanielRosenwasser) [reply src/lib/es5.d.ts]: Yeah, discussed offline, but for reasons outlined below, the fact that it's not just a simple `keyof T` or `K` (where `K extends keyof T`) means that TypeScript can't trivially look at that and say "yup, we gotta keep the modifiers around on this one".

Comment (weswigham) [reply src/lib/es5.d.ts]: Could also do
```ts
type Omit<T, K extends keyof any> = T extends unknown ? Pick<T, Exclude<keyof T, K>> : never;
```
that is to say: making the builtin `Omit` distributive. (Which'd give it an alias symbol, too) A bunch of library authors end up swapping their internal `Omit` to this once they get into situations involving unions.

Distributive conditional like this have perf issues when nested repeatedly right now, tho, sooooo.... ehhhhh?

**Comment by pirix-gh** [comment]
@DanielRosenwasser is it planned to get this smart "in" in the near future ?

**Comment by DanielRosenwasser** [comment]
I'm not sure what you're referring to - the functionality here was implemented. If you have a specific suggestion for new/different functionality in mapped types it's probably best to create a separate issue.

**Comment by DanielRosenwasser** [comment]
I'm not sure what you're referring to - the changes here were merged and will be in TypeScript 3.5. If you have a specific suggestion for new/different functionality in mapped types it's probably best to create a separate issue.

**Comment by pirix-gh** [comment]
Sorry for the confusion... So you reverted the `Omit` back to what is was because it caused the modifiers to be discarded (optional, readonly). And you explained that this is happening because the `in` keyword only maintains the modifiers with `keyof` (but not conditional types).

So my question is: **will `in` also preserve modifiers when used with conditional types in the future?**

**Comment by DanielRosenwasser** [comment]
I think it is possible, but that change is not something I can personally dedicate a ton of time towards, and there are other possible changes that are more in-demand at the moment.

**Pull request merged**